File: sdk/php/example/example_api.php
<?php

use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Generate\Account\Fee\GetSpotActualFeeReq;
use KuCoin\UniversalSDK\Generate\Futures\Market\GetKlinesReq;
use KuCoin\UniversalSDK\Generate\Service\AccountService;
use KuCoin\UniversalSDK\Generate\Service\FuturesService;
use KuCoin\UniversalSDK\Generate\Service\SpotService;
use KuCoin\UniversalSDK\Generate\Spot\Order\AddOrderSyncReq;
use KuCoin\UniversalSDK\Generate\Spot\Order\CancelOrderByOrderIdSyncReq;
use KuCoin\UniversalSDK\Generate\Spot\Order\GetOrderByOrderIdReq;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;

include '../vendor/autoload.php';


function rest_example()
{
    date_default_timezone_set('UTC');
    error_reporting(E_ALL);
    ini_set('display_errors', '1');

    $key = getenv('API_KEY') ?: '';
    $secret = getenv('API_SECRET') ?: '';
    $passphrase = getenv('API_PASSPHRASE') ?: '';

    $brokerName = getenv('BROKER_NAME');
    $brokerKey = getenv('BROKER_KEY');
    $brokerPartner = getenv('BROKER_PARTNER');

    $httpTransportOption = (new TransportOptionBuilder())
        ->setKeepAlive(true)
        ->setMaxPoolSize(10)
        ->setMaxConnectionPerPool(10)
        ->build();

    $clientOption = (new ClientOptionBuilder())
        ->setKey($key)
        ->setSecret($secret)
        ->setPassphrase($passphrase)
        ->setBrokerName($brokerName)
        ->setBrokerKey($brokerKey)
        ->setBrokerPartner($brokerPartner)
        ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
        ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
        ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
        ->setTransportOption($httpTransportOption)
        ->build();

    $client = new DefaultClient($clientOption);
    $kucoinRestService = $client->restService();

    account_service_example($kucoinRestService->getAccountService());
    spot_service_example($kucoinRestService->getSpotService());
    futures_service_example($kucoinRestService->getFuturesService());
}

function account_service_example(AccountService $accountService)
{
    $accountApi = $accountService->getAccountApi();
    $accountInfoResp = $accountApi->getAccountInfo();
    error_log("account info: level: {$accountInfoResp->level}, SubAccountSize: {$accountInfoResp->subQuantity}");

    $feeApi = $accountService->getFeeApi();
    $getActualFeeReq = GetSpotActualFeeReq::builder()
        ->setSymbols("BTC-USDT,ETH-USDT")
        ->build();

    $getActualFeeResp = $feeApi->getSpotActualFee($getActualFeeReq);

    foreach ($getActualFeeResp->data as $feeData) {
        error_log("Fee info: symbol: {$feeData->symbol}, TakerFee: {$feeData->takerFeeRate}, MakerFee: {$feeData->makerFeeRate}");
    }
}

function spot_service_example(SpotService $spotService)
{
    $orderApi = $spotService->getOrderApi();

    $addOrderReq = AddOrderSyncReq::builder()
        ->setClientOid(uniqid('', true))
        ->setSide('buy')
        ->setSymbol("BTC-USDT")
        ->setType('limit')
        ->setRemark("sdk_example")
        ->setPrice("10000")
        ->setSize("0.001")
        ->build();

    $resp = $orderApi->addOrderSync($addOrderReq);
    error_log("Add order success, id: {$resp->orderId}, oid: {$resp->clientOid}");

    $queryOrderDetailReq = GetOrderByOrderIdReq::builder()
        ->setOrderId($resp->orderId)
        ->setSymbol("BTC-USDT")
        ->build();
    $orderDetailResp = $orderApi->getOrderByOrderId($queryOrderDetailReq);
    error_log("Order detail: " . $orderDetailResp->jsonSerialize(JMS\Serializer\SerializerBuilder::create()->build()));

    $cancelOrderReq = CancelOrderByOrderIdSyncReq::builder()
        ->setOrderId($resp->orderId)
        ->setSymbol("BTC-USDT")
        ->build();
    $cancelOrderResp = $orderApi->cancelOrderByOrderIdSync($cancelOrderReq);
    error_log("Cancel order success, id: {$cancelOrderResp->orderId}");
}

function futures_service_example(FuturesService $futuresService)
{
    $marketApi = $futuresService->getMarketApi();

    $allSymbolResp = $marketApi->getAllSymbols();
    $maxQuery = min(count($allSymbolResp->data), 10);

    for ($i = 0; $i < $maxQuery; $i++) {
        $symbol = $allSymbolResp->data[$i];

        $start = (int)((microtime(true) - 600) * 1000);
        $end = (int)(microtime(true) * 1000);

        $getKlineReq = GetKlinesReq::builder()
            ->setSymbol($symbol->symbol)
            ->setGranularity(1)
            ->setFrom($start)
            ->setTo($end)
            ->build();

        $getKlineResp = $marketApi->getKlines($getKlineReq);
        $rows = [];


        foreach ($getKlineResp->data as $row) {
            $timestamp = date("Y-m-d H:i:s", $row[0] / 1000);
            $formattedRow = sprintf(
                "[Time: %s, Open: %.2f, High: %.2f, Low: %.2f, Close: %.2f, Volume: %.2f]",
                $timestamp, $row[1], $row[2], $row[3], $row[4], $row[5]
            );
            $rows[] = $formattedRow;
        }

        error_log("Symbol: {$symbol->symbol}, Kline: " . implode(', ', $rows));
    }
}

if (php_sapi_name() === 'cli') {
    rest_example();
}

File: sdk/java/example/src/main/java/com/kucoin/example/ExampleAPI.java
package com.kucoin.example;

import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.api.KucoinRestService;
import com.kucoin.universal.sdk.generate.account.account.AccountApi;
import com.kucoin.universal.sdk.generate.account.account.GetAccountInfoResp;
import com.kucoin.universal.sdk.generate.account.fee.FeeApi;
import com.kucoin.universal.sdk.generate.account.fee.GetSpotActualFeeData;
import com.kucoin.universal.sdk.generate.account.fee.GetSpotActualFeeReq;
import com.kucoin.universal.sdk.generate.account.fee.GetSpotActualFeeResp;
import com.kucoin.universal.sdk.generate.futures.market.*;
import com.kucoin.universal.sdk.generate.service.AccountService;
import com.kucoin.universal.sdk.generate.service.FuturesService;
import com.kucoin.universal.sdk.generate.service.SpotService;
import com.kucoin.universal.sdk.generate.spot.order.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class ExampleAPI {

  public static void restExample() {
    // Retrieve API secret information from environment variables
    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    // Optional: Retrieve broker secret information from environment variables; applicable for
    // broker API only
    String brokerName = System.getenv("BROKER_NAME");
    String brokerKey = System.getenv("BROKER_KEY");
    String brokerPartner = System.getenv("BROKER_PARTNER");

    // Set specific options, others will fall back to default values
    TransportOption httpTransportOption = TransportOption.builder().keepAlive(true).build();

    // Create a client using the specified options
    ClientOption clientOption =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .brokerName(brokerName)
            .brokerKey(brokerKey)
            .brokerPartner(brokerPartner)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransportOption)
            .build();

    KucoinClient client = new DefaultKucoinClient(clientOption);
    KucoinRestService kucoinRestService = client.getRestService();

    accountServiceExample(kucoinRestService.getAccountService());
    spotServiceExample(kucoinRestService.getSpotService());
    futuresServiceExample(kucoinRestService.getFuturesService());
  }

  public static void accountServiceExample(AccountService accountService) {
    AccountApi accountApi = accountService.getAccountApi();
    GetAccountInfoResp accountInfoResp = accountApi.getAccountInfo();
    log.info(
        "account info: level: {}, SubAccountSize: {}",
        accountInfoResp.getLevel(),
        accountInfoResp.getSubQuantity());

    FeeApi feeApi = accountService.getFeeApi();
    GetSpotActualFeeReq getActualFeeReq =
        GetSpotActualFeeReq.builder().symbols("BTC-USDT,ETH-USDT").build();

    GetSpotActualFeeResp getActualFeeResp = feeApi.getSpotActualFee(getActualFeeReq);

    for (GetSpotActualFeeData feeData : getActualFeeResp.getData()) {
      log.info(
          "Fee info: symbol: {}, TakerFee: {}, MakerFee: {}",
          feeData.getSymbol(),
          feeData.getTakerFeeRate(),
          feeData.getMakerFeeRate());
    }
  }

  public static void spotServiceExample(SpotService spotService) {
    OrderApi orderApi = spotService.getOrderApi();

    AddOrderSyncReq addOrderReq =
        AddOrderSyncReq.builder()
            .clientOid(UUID.randomUUID().toString())
            .side(AddOrderSyncReq.SideEnum.BUY)
            .symbol("BTC-USDT")
            .type(AddOrderSyncReq.TypeEnum.LIMIT)
            .remark("sdk_example")
            .price("10000")
            .size("0.001")
            .build();

    AddOrderSyncResp resp = orderApi.addOrderSync(addOrderReq);
    log.info("Add order success, id: {}, oid: {}", resp.getOrderId(), resp.getClientOid());

    GetOrderByOrderIdReq queryOrderDetailReq =
        GetOrderByOrderIdReq.builder().orderId(resp.getOrderId()).symbol("BTC-USDT").build();
    GetOrderByOrderIdResp orderDetailResp = orderApi.getOrderByOrderId(queryOrderDetailReq);
    log.info("Order detail: {}", orderDetailResp.toString());

    CancelOrderByOrderIdSyncReq cancelOrderReq =
        CancelOrderByOrderIdSyncReq.builder().orderId(resp.getOrderId()).symbol("BTC-USDT").build();
    CancelOrderByOrderIdSyncResp cancelOrderResp =
        orderApi.cancelOrderByOrderIdSync(cancelOrderReq);
    log.info("Cancel order success, id: {}", cancelOrderResp.getOrderId());
  }

  public static void futuresServiceExample(FuturesService futuresService) {
    MarketApi marketApi = futuresService.getMarketApi();

    GetAllSymbolsResp allSymbolResp = marketApi.getAllSymbols();
    int maxQuery = Math.min(allSymbolResp.getData().size(), 10);

    for (int i = 0; i < maxQuery; i++) {
      GetAllSymbolsData symbol = allSymbolResp.getData().get(i);

      long start = (long) ((System.currentTimeMillis() - 600000));
      long end = (long) (System.currentTimeMillis());

      GetKlinesReq getKlineReq =
          GetKlinesReq.builder()
              .symbol(symbol.getSymbol())
              .granularity(GetKlinesReq.GranularityEnum._1)
              .from(start)
              .to(end)
              .build();

      GetKlinesResp getKlineResp = marketApi.getKlines(getKlineReq);
      List<String> rows = new ArrayList<>();

      for (List<Double> row : getKlineResp.getData()) {
        String timestamp =
            new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date(row.get(0).longValue()));
        String formattedRow =
            String.format(
                "[Time: %s, Open: %.2f, High: %.2f, Low: %.2f, Close: %.2f, Volume: %.2f]",
                timestamp, row.get(1), row.get(2), row.get(3), row.get(4), row.get(5));
        rows.add(formattedRow);
      }

      log.info("Symbol: {}, Kline: {}", symbol.getSymbol(), String.join(", ", rows));
    }
  }

  public static void main(String[] args) {
    restExample();
  }
}
